src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs

```diff
@@ -1863,6 +1867,11 @@ private EventBuilder DefineEventNoLock(string name, EventAttributes attributes,
 
             tkType = m_module.GetTypeTokenInternal(eventtype);
 
+            // For compat, check for ByRef after acquiring the token. The primary
+            // compat issue here is throwing on a null instance.
```

Comment:
Where is the problem with throwing on a null instance? I do not see it.

Reply:
There is already a check for `null` of `eventtype` in `GetTypeTokenInternal` so I am relying on that. I could instead throw the correct exception here too but decided to leave it logically as desired since `GetTypeTokenInternal` has throws the exception in many other places.

Reply:
Ah ok, I understand the problem with `null`.

For the throwing on `eventtype.IsByRef`, I think we should depend on the type loader.

Reply:
> I think we should depend on the type loader.

I agree in principle but that is a far broader change in this case. It would require ensure the type loader throws the correct exception and has the opportunity to do so at the same time. Based on what I am seeing in the Reflection this isn't always possible. The state isn't always communicated at the correct time.

Reply:
My goal for much of this is that the current testing remains unchanged. I've tried to move a few things and many tests start failing because of when they expect an exception to occur.

Reply:
Reflection.Emit expects you to know what you are doing and emit correct IL and metadata. It does not validate whether the IL and metadata is valid. Trying to duplicate checks done by the type loader in Reflection.Emit often got us in trouble since these checks got out of sync with type loader.

For example, I believe that there are other cases that are invalid for event and that could be detected here, but we are not doing that.

> I've tried to move a few things and many tests start failing because of when they expect an exception to occur.

What are the tests that started failing? We should update them.

Reply:
>We should update them.

Agreed, but not in this PR. I don't think that updates to the Reflection API and its exceptions semantics should be changed here because this PR is about adding support in the runtimes and not comprehensive support in the Reflection API.

I will remove them again and record the failures.